File: snakes-and-ladders.js
Time: O(M*N)
Space: O(M*N)

/**
 * @param {number[][]} board
 * @return {number}
 */
var snakesAndLadders = function(board) {
    if (!board || !board.length) return 0;
    
    const N = board.length**2,
          flattenBoard = new Array(N);
    let leftToRight = true,
        i = board.length - 1,
        j = 0,
        index = 0;
    
    while (i >= 0 && j >= 0) {
        if (board[i][j] == -1) flattenBoard[index++] = -1;
        else flattenBoard[index++] = board[i][j] - 1;
        
        if (leftToRight) j++;
        else j--;
        if (j >= board.length) {
            j--;
            i--;
            leftToRight = false;
        } else if (j < 0) {
            j++;
            i--;
            leftToRight = true;
        }
    }

    const q = [];
    let steps = 0;
    q.push(0);
    flattenBoard[0] = -2;

    while (q.length > 0) {
        let level = q.length;
        
        for (i = 0; i < level; i++) {
            let current = q.shift();
            console.log(current);
            if (current == N - 1) return steps;
            
            for (j = 1; j < 7; j++) {
                let child = current + j;

                if (child < N && flattenBoard[child] != -2) {
                    if (flattenBoard[child] >= 0)
                        q.push(flattenBoard[child]);
                    else
                        q.push(child);

                    flattenBoard[child] = -2;
                }

            }
        }
        steps++;
    }
    
    return -1;
};

File: problem2/SnakesAndLadders.java
// Time Complexity : O(n*n) n -> Length of board
// Space Complexity : O(n*n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
package problem2;

import java.util.LinkedList;
import java.util.Queue;

public class SnakesAndLadders {
	public int snakesAndLadders(int[][] board) {
		if (board == null || board.length == 0) {
			return 0;
		}

		int[] moves = flattenBoard(board);
		Queue<Integer> queue = new LinkedList<>();
		queue.add(1);
		moves[1] = -2;

		int n = board.length;
		int min = 0;

		while (!queue.isEmpty()) {
			int size = queue.size();
			for (int i = 0; i < size; i++) {
				int curr = queue.poll();
				if (curr == n * n) {
					return min;
				}

				for (int j = 1; j < 7; j++) {
					int child = curr + j;
					if (child > n * n) {
						break;
					}
					if (moves[child] != -2) {
						if (moves[child] == -1) {
							queue.add(child);
						} else {
							queue.add(moves[child]);
						}
						moves[child] = -2;
					}
				}
			}
			min++;
		}

		return -1;

	}

	private int[] flattenBoard(int[][] board) {
		int n = board.length;
		int[] moves = new int[n * n + 1];

		int i = n - 1;
		int j = 0;
		int idx = 1;
		int even = 0;

		while (i >= 0 && j >= 0) {
			moves[idx] = board[i][j];
			idx++;

			if (even % 2 == 0) {
				j++;
				if (j == n) {
					j--;
					i--;
					even++;
				}
			} else {
				j--;
				if (j == -1) {
					j++;
					i--;
					even++;
				}
			}
		}

		return moves;
	}

	public static void main(String[] args) {
		SnakesAndLadders obj = new SnakesAndLadders();
		int[][] board = { { -1, -1, -1, -1, -1, -1 }, { -1, -1, -1, -1, -1, -1 }, { -1, -1, -1, -1, -1, -1 },
				{ -1, 35, -1, -1, 13, -1 }, { -1, -1, -1, -1, -1, -1 }, { -1, 15, -1, -1, -1, -1 } };

		System.out.println("Minimum number of moves to reach destination: " + obj.snakesAndLadders(board));
	}

}
